### PR TITLE
site/retirements: download certificates

### LINKED ISSUE #531: Create KlimaDAO branded PDF receipts for retirements
# Summary
Customers are asking for old-school PDF certificates similar to those created by the Verra registry.
The original plan was to get these from the underlying bridge (c3 or toucan). But we don't want to wait anymore.
Now we think it makes sense to generate a KlimaDAO specific one.

# PDF Requirements
See design https://www.figma.com/file/RtYPQewQkz4RENj81ugUnc/PDF-Certificate?node-id=0%3A1

# Technical Requirements
- Nextjs API endpoint that accepts `address` and `index` and responds with a PDF download
- Cache response for 1 day
- Use a lightweight JavaScript PDF generator library for server-side pdf generation, like `JSPDF` or `PDFMake` or `PDFKit`
- "Download PDF" button on all retirement pages (where the "coming soon" button is)

### PR DESCRIPTION
**Pull request by 0xAeterno**
## Description

Generate retirement certificates via the retirement page.

Note:
- `Type` and `NFT Certification ID` attributes are pending data from an updated subgraph deployment

### TODO:
- [x] Handle Moss certificate content (need background asset)
- [x] Add link to projects that back mco2 pools for moss certs (verra project details arent currently available for moss projects) 
- [x] Refactor code

### Preview Link
[Mark Cuban's retirements](https://klimadao-site-git-retirement-certificates-klimadao.vercel.app/retirements/markcuban.klima)
[Polygon's retirements](https://klimadao-site-git-retirement-certificates-klimadao.vercel.app/retirements/0x087a7afb6975a2837453be685eb6272576c0bc06)
[Address with various token retirements](https://klimadao-site-git-retirement-certificates-klimadao.vercel.app/retirements/0xddff75a29eb4bfecf65380de9a75ad08c140ea49)
## Related Ticket

Resolves #531 

## Changes

| Token  | Certificate  |
|---------|--------|
| BCT |image|
| NCT |image|
| NBO |image|
| UBO |image|
| MCO2 |image|

## Checklist

- [x] Building the site with `npm run build-site` works without errors
- [x] Building the app with `npm run build-app` works without errors
- [x] I formatted JS and TS files with running `npm run format-all`

### CODE REVIEW

File: site/components/pages/Retirements/SingleRetirement/ProjectDetails/List/index.tsx [hunk 1]
```diff
@@ -18,6 +18,8 @@ export const ProjectDetail: FC<Props> = (props) => {
   const { projectLink, headline, tokenAddress, totalRetired } = props;
   const { locale } = useRouter();
 
+  console.log(props);
+
```
Comment (0xAeterno) [inline site/components/pages/Retirements/SingleRetirement/ProjectDetails/List/index.tsx:22]:
```suggestion
```

**Comment by 0xAeterno** [comment]
@Atmosfearful this is ready for review and testing with the product team.

A couple things:
- `jspdf` is a client side pdf generation library, therefore having the assets required does increase the bundle size a bit. We lazily load the pdf download button on the single retirement page to mitigate some of this.
- Dynamically setting a different background for each token does contribute to the bundle size too, I've optimised and reduced the images as much as I can without sacrificing quality.

Keen to hear your thoughts, I'm pretty happy with the design we've landed on for the certificates :)

File: site/components/pages/Retirements/SingleRetirement/DownloadCertificateButton/generateCertificate.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  setupFonts();
+  printHeader();
+  printFeatureImage();
+  printRetirementDetails();
+  printTransactionDetails();
+  printProjectDetails();
+  if (isMossRetirement) printMossProjectDetails();
+  printRetirementLink();
+
+  doc.save(`${params.retirement.id}.pdf`);
```
Comment (0xAeterno) [inline site/components/pages/Retirements/SingleRetirement/DownloadCertificateButton/generateCertificate.ts:279]: Is there a more meaningful file name we can save the certificate name as?

**Comment by jabby09** [comment]
@Atmosfearful @0xAeterno this works for me on desktop and mobile. It does actually download the PDF right away, vs open it in your browser (as a preview and then download) but I assume that is intentional.

**Comment by 0xAeterno** [comment]
@jabby09 yep the download prompt is intentional. Would we prefer to open in the browser in a new tab?

**Comment by jabby09** [comment]
I think preview first in a new tab is preferred.

File: site/components/pages/Retirements/SingleRetirement/DownloadCertificateButton/generateCertificate.ts [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    klimaLogo.src = KlimaLogo.src;
+    doc.addImage(klimaLogo, "JPEG", spacing.margin, spacing.margin, 60, 10);
+
+    doc.setFont("Poppins", "Bold");
+    doc.setFontSize(24);
+    doc.text("Certificate for On-chain", spacing.margin, 36);
+    doc.text("Carbon Retirement", spacing.margin, 46);
+    doc.setLineWidth(1.05);
+    doc.setDrawColor(KLIMA_GREEN);
+    doc.line(spacing.margin, 52, 173, 52);
```
Comment (Atmosfearful) [inline site/components/pages/Retirements/SingleRetirement/DownloadCertificateButton/generateCertificate.ts:83]: I imagine all of this was a ton of trial and error- really appreciate it! It looks great!

File: site/components/pages/Retirements/SingleRetirement/DownloadCertificateButton/generateCertificate.ts (hunk 2 above)
Comment (Atmosfearful) [reply site/components/pages/Retirements/SingleRetirement/DownloadCertificateButton/generateCertificate.ts:280]: retirement_01_0x12345678912381234567.pdf
`retirement_{INDEX}_{BENEFICIARY_ADDRESS}.pdf`

**Review by Atmosfearful — APPROVED** [review]
Testing:
before - old page - `fast 3G, disable cache, hard refresh`
image

after - `fast 3G, disable cache, hard refresh`
image

So about a 20% slowdown before the `onLoad` event fires. This layout shift of the lazyloaded button also hurts our 'experience' score, and of course the original page is a bit bloated even before this PR.

Let's go ahead and invest the time in moving this into a nextjs api endpoint. Should not be too difficult, it looks like this code will work fine in a node environment!

This is fine for production though!

**Pull request merged**